lib/Db/MessageMapper.php

```diff
@@ -924,25 +955,30 @@ public function findRelatedData(array $messages, string $userId): array {
 	 * @return int[]
 	 */
 	public function findNewIds(Mailbox $mailbox, array $ids): array {
-		$qb = $this->db->getQueryBuilder();
 		$sub = $this->db->getQueryBuilder();
+		$qb = $this->db->getQueryBuilder();
 
 		$subSelect = $sub
 			->select($sub->func()->max('uid'))
 			->from($this->getTableName())
 			->where(
 				$sub->expr()->eq('mailbox_id', $qb->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT)),
-				$sub->expr()->in('id', $qb->createNamedParameter($ids, IQueryBuilder::PARAM_INT_ARRAY), IQueryBuilder::PARAM_INT_ARRAY)
+				$sub->expr()->in('id', $qb->createParameter('ids'))
 			);
-		$select = $qb
-			->select('id')
+
+		$qb->select('id')
 			->from($this->getTableName())
 			->where(
-				$qb->expr()->eq('mailbox_id', $qb->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT)),
-				$qb->expr()->gt('uid', $qb->createFunction('(' . $subSelect->getSQL() . ')'), IQueryBuilder::PARAM_INT)
+				$qb->expr()->eq('mailbox_id', $qb->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT))
 			);
 
-		return $this->findIds($select);
+		return array_flat_map(function (array $chunk) use ($qb, $subSelect) {
+			$qb->setParameter('ids', $chunk, IQueryBuilder::PARAM_INT_ARRAY);
+			$select = $qb->where(
```

Comment:
Caution: `where` is going to overwrite a previous value. 

```
$qb->select('id')
	->from($this->getTableName())
	->where(
		$qb->expr()->eq('mailbox_id', $qb->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT))
	);
```

```
SELECT `id` FROM `*PREFIX*mail_messages` WHERE `mailbox_id` = :dcValue2
```

```
$qb->setParameter('ids', $chunk, IQueryBuilder::PARAM_INT_ARRAY);
$select = $qb->where(
	$qb->expr()->gt('uid', $qb->createFunction('(' . $subSelect->getSQL() . ')'), IQueryBuilder::PARAM_INT)
);
```
```
SELECT `id` FROM `*PREFIX*mail_messages` WHERE `uid` > (SELECT MAX(`uid`) FROM `*PREFIX*mail_messages` WHERE (`mailbox_id` = :dcValue1) AND (`id` IN (:ids)))
```

The comparison for mailbox_id in the outer query is lost. I guess the subselect is always the same and could be moved outside array_flat_map. To append another expression to an existing where clause: `andWhere` / `orWhere`.

Reply:
No, the subselect can't be moved outside. The parameter `ids` is set on `$qb` inside `array_flat_map` to chunk the ids for Oracle DBs.